examples/Conversational_agents_using_tools.ipynb

```diff
@@ -0,0 +1,1095 @@
+    "    # If you change your prompt template you'll need to adjust this as well\n",
+    "    stop=[\"\\nObservation:\"], \n",
+    "    allowed_tools=tool_names\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "dd893fa5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Initiate the agent that will respond to our queries\n",
+    "# Set verbose=True to share the CoT reasoning the LLM goes through\n",
+    "agent_executor = AgentExecutor.from_agent_and_tools(agent=agent, tools=tools, verbose=True)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "c7ca3ee2",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n",
+      "\n",
+      "\u001B[1m> Entering new AgentExecutor chain...\u001B[0m\n",
+      "\u001B[32;1m\u001B[1;3mThought: Hmm, I be not sure of the answer to that one. Let me think.\n",
+      "Action: Search\n",
+      "Action Input: \"Canada population 2023\"\u001B[0m\n",
+      "\n",
+      "Observation:\u001B[36;1m\u001B[1;3mThe current population of Canada is 38,662,830 as of Monday, April 17, 2023, based on Worldometer elaboration of the latest United Nations data.\u001B[0m\u001B[32;1m\u001B[1;3mAhoy, that be the answer I was lookin' for!\n",
+      "Final Answer: The population of Canada as of 2023 be 38,662,830, Arg!\u001B[0m\n",
+      "\n",
+      "\u001B[1m> Finished chain.\u001B[0m\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "'The population of Canada as of 2023 be 38,662,830, Arg!'"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "agent_executor.run(\"How many people live in canada as of 2023?\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c3b51817",
+   "metadata": {},
+   "source": [
+    "## LLM Agent with History\n",
+    "\n",
+    "Extend the LLM Agent with the ability to retain a [memory](https://python.langchain.com/en/latest/modules/agents/agents/custom_llm_agent.html#adding-memory) and use it as context as it continues the conversation.\n",
+    "\n",
+    "We use a simple ConversationBufferWindowMemory for this example that keeps a rolling window of the last two conversation turns, but LangChain have many other [memory options](https://python.langchain.com/en/latest/modules/memory.html) depending on your use case."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "id": "85737eea",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Set up a prompt template which can interpolate the history\n",
+    "template_with_history = \"\"\"You are SearchGPT, a professional search engine who aims to provide informative answers to users. Answer the following questions as best you can. You have access to the following tools:\n",
```

Comment:
suggestion: In my prompts I prefer to say "who provides informative answers to users" over "who aims to to provide informative answers to users." I haven't tested it, but my theory is that when you say the model tries (rather than succeeds), it makes the model more likely to generate completions that fail.

Reply:
Makes sense, I'll make this change and keep in mind for future